feat: add offset value for showing header smoothly

david50407 opened the pull request:
For now, `showHeader()` changes `positionY` directly, this causes header flashes back its hidden state while scrolling right after `showHeader()` because `onScroll` event will override this value from far.

I'm introducing a new attribute `offsetY`, this value is not affected by `onScroll` and can offset position.
We can simply show header by subtracting `headerHeight` from `offsetY`, and `offsetY` also can be animated.

benevbright commented on src/core.tsx:
```diff
@@ -91,17 +92,25 @@ const useCollapsibleHeader = (
+  const offsetY = React.useRef(new Animated.Value(0)).current;
+  const offsetYValue = React.useRef(0);
+  React.useEffect(() => {
+    const listener = offsetY.addListener(({ value }) => {
```
what is this `offsetY.addListener`? Could you explain a bit about It's listening what?

david50407 replied:
I save current value of `offsetY` into `offsetYValue`. The listener here will receive value of `offsetY` during the transition.

benevbright reviewed (commented):
Amazing! I tested and it's working well. But I don't know how does it work 😳 and how its animation is working.. Could you explain a bit? 

And please fix the simple linter issue. try to run `yarn lint` please.
Thanks again.

david50407 commented:
I changed `animatedDiffClampY` from only `positionY` to `positionY + offsetY`.

The main reason is `positionY` will be overwritten by `onScroll` event to where the list scrolled.
In the original way, if we just reset `positionY` to `0` directly, the header will show. But on the next scroll, `positionY` will be assigned by a far value, this cause the header hide immediately again.

So I introduced the value `offsetY` to give a offset on positionY which won't be affected by `onScroll` event, if we want to show header, just set `offsetY` to `offsetY - headerHeight`, this will cheat `animatedDiffClampY` like we scrolled down a distence of header height, then the header shows. And because `offsetY` is still at scrollable's current position, in the next scroll, the change of `offsetY` won't be large.

The idea is inspired from [this post](https://medium.com/appandflow/react-native-collapsible-navbar-e51a049b560a), I just borrowed the offset part from there.

If you want more smoothly animation while showing header, just replace `showHeader` method with:
```ts
Animated.timing(offsetY, {
  toValue,
  duration: 350,
  useNativeDriver: true,
}).start();
```

And that's why I also expose `offsetY` that can be used to create any animation.

david50407 commented:
btw, I do not perform a check that value of `offsetY` will be overflowed or not.

If you're considering the overflow problem, we can reset `offsetY` into a safe-range sliently when header is fully shown or hidden.

benevbright commented on src/core.tsx:
```diff
@@ -184,10 +196,13 @@ const useCollapsibleHeader = (
       containerPaddingTop: 0,
       scrollIndicatorInsetTop: 0,
       positionY: new Animated.Value(0),
+      offsetY: new Animated.Value(0),
```
Any reason to give an opportunity to use for the dev?

benevbright replied:
> And that's why I also expose offsetY that can be used to create any animation.

Sorry, I was too early. Thanks for the explanation.

benevbright commented:
@david50407 Hey, David. Great. Thanks again for the PR and the explanation. 👍
